ethcore/src/engines/authority_round/mod.rs

```diff
@@ -125,54 +140,96 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
 	}
 }
 
-// Helper for managing the step.
+/// Helper for managing the step.
 #[derive(Debug)]
 struct Step {
 	calibrate: bool, // whether calibration is enabled.
-	inner: AtomicUsize,
-	duration: u16,
+	inner: AtomicU64,
+	/// Planned durations of steps.
+	durations: BTreeMap<u64, u16>,
 }
 
 impl Step {
-	fn load(&self) -> u64 { self.inner.load(AtomicOrdering::SeqCst) as u64 }
+	fn load(&self) -> u64 { self.inner.load(AtomicOrdering::SeqCst) }
+
 	fn duration_remaining(&self) -> Duration {
+		self.opt_duration_remaining().unwrap_or_else(|| {
+			let ctr = self.load();
+			error!(target: "engine", "Step counter under- or overflow: {}, aborting", ctr);
+			panic!("step counter under- or overflow: {}", ctr)
+		})
+	}
+
+	fn opt_duration_remaining(&self) -> Option<Duration> {
 		let now = unix_now();
-		let expected_seconds = self.load()
-			.checked_add(1)
-			.and_then(|ctr| ctr.checked_mul(self.duration as u64))
-			.map(Duration::from_secs);
-
-		match expected_seconds {
-			Some(step_end) if step_end > now => step_end - now,
-			Some(_) => Duration::from_secs(0),
-			None => {
-				let ctr = self.load();
-				error!(target: "engine", "Step counter is too high: {}, aborting", ctr);
-				panic!("step counter is too high: {}", ctr)
-			},
+		let mut prev_dur = u64::from(self.durations[&0]);
+		let mut prev_step = 0u64;
+		let mut prev_time = 0u64;
+		let next_step = self.load().checked_add(1)?;
+		for (time, dur) in self.durations.iter().skip(1) {
```

Comment:
Is this loop supposed to find the transition timestamp and duration of current step? If so I think it could be greately simplified (and optimized) by:
1. Converting timestamps into step transitions + keeping the timestamps
2. Bin searching in the step transitions to find current which transition we fall into and just use the pre-computed starting timestamp.

Instead of bin-searching I think we should rather start from the end, as that's a most likely case that we will fall into.
So to summarize, we need:
`Vec<(StepTransition, TimestampTransition, StepDuration)>`
Then you start from the end and check if `next_step > StepTransition` if it is you use `TimestampTransition + (next_step - StepTransition) * StepDuration` to compute the expected timestamp of the next step.